### PR TITLE
Fix Z and M coordinates edition. Add string to double conversion from…

### PR DESCRIPTION
**Pull request by ghost**
Fix Z & M values edition. Associated QgsAbstractGeometry doesn't have Z/M support. Support will be enabled on the fly (not sure this is the correct way).

String to double conversion uses QLocale::toDouble() instead of QString::toDouble() ; to support conversion for numbers with comma decimal separator.

### CODE REVIEW

File: src/app/nodetool/qgsnodeeditor.cpp [hunk 1]
```diff
@@ -200,15 +200,18 @@ bool QgsNodeEditorModel::setData( const QModelIndex &index, const QVariant &valu
...
   }
 
-  double x = ( index.column() == 0 ? value.toDouble() : mSelectedFeature->vertexMap().at( index.row() )->point().x() );
-  double y = ( index.column() == 1 ? value.toDouble() : mSelectedFeature->vertexMap().at( index.row() )->point().y() );
+  // Get double value wrt current locale.
+  QLocale locale;
+  double doubleValue = locale.toDouble( value.toString() );
+
+  double x = ( index.column() == 0 ? doubleValue : mSelectedFeature->vertexMap().at( index.row() )->point().x() );
+  double y = ( index.column() == 1 ? doubleValue : mSelectedFeature->vertexMap().at( index.row() )->point().y() );
```
Comment (vmora) [inline src/app/nodetool/qgsnodeeditor.cpp]: I suggest::
```
const double doubleValue = QLocale().toDouble( value.toString() );
```

File: src/app/qgsmaptooladdfeature.cpp [hunk 2]
```diff
@@ -175,6 +175,14 @@ void QgsMapToolAddFeature::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
         g = QgsGeometry::fromPoint( savePoint );
       }
 
+      // Use default Z-value if coordinates is enabled.
+      if (QgsWkbTypes::hasZ(layerWKBType))
+          g.geometry()->addZValue(defaultZValue());
+
+      // Use default M-value if coordinates is enabled.
+      if (QgsWkbTypes::hasM(layerWKBType))
+          g.geometry()->addMValue(0.0 /* No default M-value yet. */);
```
Comment (vmora) [inline src/app/qgsmaptooladdfeature.cpp]: wouldn't be better to define `defaultMValue()` ?

File: src/core/qgsvectorlayereditutils.cpp [hunk 3]
```diff
@@ -84,6 +84,16 @@ bool QgsVectorLayerEditUtils::moveVertex( const QgsPoint &p, QgsFeatureId atFeat
 
   QgsGeometry geometry = f.geometry();
 
+  QgsAbstractGeometry *abstractGeometry = geometry.geometry();
+
+  // Support Z value.
+  if ( abstractGeometry && QgsWkbTypes::hasZ( p.wkbType() ) )
+    abstractGeometry->addZValue( 0.0 );
```
Comment (vmora) [inline src/core/qgsvectorlayereditutils.cpp]: is the defaultZValue unusable here ?

**Comment by vmora** [comment]
also, you should run `scripts/prepare-commit.sh` for indentation, or even better `cd .git/hooks/; ln -s ../../scripts/prepare-commit.sh pre-commit`

File: src/app/qgsmaptooladdfeature.cpp (hunk 2 above)
Comment (m-kuhn) [reply src/app/qgsmaptooladdfeature.cpp]: Better use `std::numeric_limits<double>::quiet_NaN()`

**Comment by nouanda** [comment]
Default Z value has been moved from QGis global properties to QGis project properties.

**Comment by nyalldawson** [comment]
There's lots of discussion relating to default z value handling in https://github.com/qgis/QGIS/pull/3905. I believe it was by design that it was a global option. @haubourg may be able to clarify here.

**Comment by haubourg** [comment]
@nyalldawson Yes, after some testing, and discussion with @3nids we found that this setting had a lot more meaning being in the project properties. Sorry for the ping pong party :)

**Comment by nyalldawson** [comment]
That's fine - just wanted to make sure this wasn't clashing with your earlier work

**Comment by pblottiere** [comment]
@nouanda I would like to test this PR. Do you plan to rebase / continue this work?

**Comment by gioman** [comment]
Is this somehow related to this issues I'm seeing in 2.18? https://issues.qgis.org/issues/17247

**Comment by nouanda** [comment]
Clean PR: https://github.com/qgis/QGIS/pull/5402

**Pull request closed without merging**